packages/mermaid/src/diagrams/info/infoParser.ts

```diff
@@ -0,0 +1,20 @@
+import { EmptyFileSystem, type LangiumParser, type ParseResult } from 'langium';
+import { type Info, type InfoServices, createInfoServices } from 'mermaid-parser';
+
+import { log } from '../../logger.js';
+
+const services: InfoServices = createInfoServices(EmptyFileSystem).Info;
+const langiumParser: LangiumParser = services.parser.LangiumParser;
+
+export function parse(input: string): void {
+  const result: ParseResult<Info> = langiumParser.parse<Info>(input);
+  if (result.parserErrors.length > 0 || result.lexerErrors.length > 0) {
+    const parserErrors = result.parserErrors.map((parseError) => parseError);
+    const lexerErrors = result.lexerErrors.map((lexerError) => lexerError);
+    log.error(
+      { parserErrors: parserErrors, lexerErrors: lexerErrors },
+      'Error parsing info diagram'
+    );
+    throw new Error(`Parser errors: ${parserErrors}\nLex errors: ${lexerErrors}`);
+  }
+}
```

Comment:
I think we should remove all these from `mermaid` package, and encapsulate all langium related logic inside the other package. 

```mermaid
flowchart TD
    JISON["Current"] --> Text["Text"] --> B(Loop through detectors)
    B --> C{Does it match?}
    C -->|No| B
    C -->|Yes| E("new Diagram()")
    E --> P(Parse)
    P --> DB(Update DB)
    DB --> R(Render)

    Langium --> Text2["Text"] --> Par("const {type, ast} = mermaid-parser.parse(text)")
    Par --> ND("new Diagram({type, ast})")
    ND --> PDB("populateDB(ast)")
    PDB --> R
```

Reply:
So, the only things exported from the new package should be the `parse` function and the AST types.

Reply:
It make sense to have the logic in the other package, then I'll export that `parse` function, but how to handle errors? I don't think we should make it throw an error it has lexer or parse errors.

And also should we remove other attributes in the AST? for example `endColumn` and `startColumn` of each field ...etc.

But seeing that removed the detector in langium path then I'm assuming you're referring to `MermaidRegistryServices`. I'm sure how it's works yet, so we still need the more modification and testing to make its working as expected.

Reply:
> I don't think we should make it throw an error it has lexer or parse errors.

Why not?

We can use a try catch when calling `parse` to handle those errors no?

Reply:
The detectors don't necessarily need to go away, we can use that inside the parser. 

Also, the lazy loading is something we need to keep in mind. Although not the priority now.

Reply:
> And also should we remove other attributes in the AST? for example endColumn and startColumn of each field ...etc.

What all info is there in the AST now?

Reply:
> Why not?
> 
> We can use a try catch when calling `parse` to handle those errors no?

Not sure why :)

> The detectors don't necessarily need to go away, we can use that inside the parser.


I think the `MermaidRegistryServices` is doing the same thing here, but not sure how to use it yet.

> Also, the lazy loading is something we need to keep in mind. Although not the priority now.

All services are already lazy loaded.

> What all info is there in the AST now?

Actual value (with other values):

```ts
interface Info {
  $type: 'Info';
  title?: string;
  accTitle?: string;
  accDescr?: string;
  version?: string;
}
```

You might ask why we need title and accessibilities, langium entry rule must have an assignment, so we can't have empty AST, that's why I added those.

Reply:
> Not sure why :)

Because that's how errors should generally be handled. 
Pseudo code below. 

```ts
// somewhere inside mermaid
import { parse } from 'mermaid-parser';
const parseText = (text: string) => {
  try {
    const { ast, type } = await parse(text);
//    log(Got type with ast);
    return { ast, type };
  } catch (error) {
// Type check here. 
    const { parserErrors, lexerErrors } = error.details;
    log.error(
      error.details,
      'Error parsing diagram'
    );
    throw new MermaidParsingError(parserErrors, lexerErrors);
  }
}

Reply:
> All services are already lazy loaded.

Are they in different chunks now to be downloaded when necessary?

Reply:
@Yokozuna59 to make your life easier (for now), we can do this.

- Don't change existing detection logic
- Make the parse in `parser` accept the type of diagram. 
- Use switch to call necessary service inside the parse function.

Reply:
> Because that's how errors should generally be handled. (Unless there is a need to return without throwing)

Okay.

---

> Are they in different chunks now to be downloaded when necessary?

> Additionally, all services within the DI container are lazy loaded, so there's no runtime overhead unless you actually call these services.
> https://github.com/mermaid-js/mermaid/pull/4450#issuecomment-1586103778

> If you take a look at the code you will see that it only generates a small reflection layer for your grammar types. The actual grammars are stored elsewhere and are lazy loaded.
> https://github.com/mermaid-js/mermaid/pull/4450#issuecomment-1642441133

---

> @Yokozuna59 to make your life easier (for now), we can do this.
> 
>     * Don't change existing detection logic
> 
>     * Make the parse in `parser` accept the type of diagram, along with text. `parse('info', text)`
> 
>     * Use switch to call necessary service inside the parse function.

I'll ask langium team about that service to the best approach here.